### ISSUE TITLE
OnCloseRequested as a function that accepts the current model

### ISSUE THREAD

**Issue by bender2k14** [elmish/Elmish.WPF#243]
In all of the ways to create a `SubModelWin` binding, there is an optional argument of type `'msg` called `onCloseRequested`.  Here is one example.

https://github.com/elmish/Elmish.WPF/blob/783eaa42bcbd261960147cfa33ddb6f6beda9613/src/Elmish.WPF/Binding.fs#L1443

Eventually the data provided by as that argument is stored in a field called `OnCloseRequested` with type `'msg voption`

https://github.com/elmish/Elmish.WPF/blob/783eaa42bcbd261960147cfa33ddb6f6beda9613/src/Elmish.WPF/Binding.fs#L100

I would like to improve the expressiveness of the `SubModelWin` binding by changing the type of the `OnCloseRequested` field from
- from `'msg voption`
- to `'model -> 'msg voption`.

The change I want does not require a change to the public API.  However, we could expose this increased flexibility with additional overloads of `subModelWin` that have type `'model -> 'msg` for the optional argument named `onCloseRequested`.

**Comment by cmeeren** [comment]
Is that needed? You have the model when you handle the specified message in the `update` function. I can't see a way to have it both places.

**Comment by bender2k14** [comment]
The good reason is so I can implement `mapMsgWithModel`.  After adding this `'model` parameter to [`OnCloseRequested`](https://github.com/bender2k14/Elmish.WPF/commit/bf68353ee28c66cc2cf19797f38cd6f7ba74c528) and [`ToMsg`](https://github.com/bender2k14/Elmish.WPF/commit/3892323d834332cf50e516ceba7ec8e525fc272d) in those commits, I was able to add `mapMsgWithModel` in [this commit](https://github.com/elmish/Elmish.WPF/pull/244/commits/c9a9b16af6bbd2b04eca6c67b64c8251b66b46cc).

If the `'model` is now available in two places and we only want to have one, then my current feeling would be that the way I just added is the one to keep.

**Comment by bender2k14** [comment]
Oh, I think my previous commit isn't so correct.  Let me try again.

I think the best argument for adding this `'model` argument is to avoid race conditions.  When `onCloseRequested` should be executed, the `'model` instance I am suggesting be provided to it might be a different than the one provided to `update`.

Also, most other `'msg`-producing functions are already supplied to the current `'model`.  One example is the `Set` function in the two-way binding.  (The only `'msg`-producing functions that don't accept `'model` are the `OnCloseRequested` and `ToMsg` functions.)

https://github.com/elmish/Elmish.WPF/blob/c20e323c466140b14acaff9d6c531f31e5bacd46/src/Elmish.WPF/Binding.fs#L54-L58

It currently seems like a convenient coincidence to me that I also want `OnCloseRequested` (at least internally) to accept the current `'model` (in order to implement `mapMsgWithModel`).

**Comment by cmeeren** [comment]
I'm starting to come around, but I want to understand the motivation a bit better.

> I think the best argument for adding this `'model` argument is to avoid race conditions. When `onCloseRequested` should be executed, the `'model` instance I am suggesting be provided to it might be a different than the one provided to `update`.

I don't see a problem in practice. Can you convince me that if these are different, the model in `onCloseRequested` may be the correct one? I can't convince myself of that. If anything, the `update` one seems the most foolproof to me. When a close is requested, a message is dispatched to `update` regardless of the model, and how that message is handled depends on `update` and, by extension, its model. It's more logic in `update` and less in `bindings`, which IMHO is a good thing.



> Also, most other `'msg`-producing functions are already supplied to the current `'model`.

Consistency is good, but not if it isn't needed and confuses users. Given my comment above, I myself would be confused about whether to use the model in `onCloseRequested` or `update`.

**Comment by bender2k14** [comment]
> Even if we implement this as you want, we may need to document the distinction.

To repeat, I don't care very much if we add/change the public API related to the argument `onCloseRequested`.  What I care about is implementing `mapMsgWithModel`, and I was able to achieve that by changing the type of the field `OnCloseRequested` from `'msg voption` to `'model -> 'msg voption`, which leaves the public API unchanged.

The purpose of this issue was originally to point out that we could choose to take advantage of this increase in expressiveness by exposing it in the public API in a backwards-compatible way (by adding more overloads of the method `Binding.subModelWin`.  After creating this issue, I realized that nearly all of the other `'msg`-producing functions given to bindings already support this feature.

For the purposes of having a debate, I don't mind taking the position that we should expose this additional capability.  Indeed, if I had to choose, then I would make this API change.

>  Can you convince me that if these are different, the model in `onCloseRequested` may be the correct one?

Maybe, but not right now when I am so sleepy 😛   I will reply to this in the morning.
